crates/pallets/starknet/src/lib.rs

```diff
@@ -659,6 +650,17 @@ pub mod pallet {
             let chain_id = Self::chain_id();
             let transaction = input_transaction.into_executable::<T::SystemHash>(chain_id, paid_fee_on_l1, false);
 
+            let tx_hash = transaction.tx_hash;
+            let nonce: Nonce = transaction.tx.nonce;
+
+            // Ensure that L1 Message has not been executed
+            Self::ensure_l1_message_not_executed(&nonce).map_err(|_| Error::<T>::L1MessageAlreadyExecuted)?;
+
+            // Store infornamtion about message being processed
+            // The next instruction executes the message
+            // Either successfully  or not
+            L1Messages::<T>::mutate(|nonces| nonces.insert(nonce));
```

Comment:
Aren't there scenarios where the execution fails for different reasons ? Like the sequencer does not have enough ETH to pay the fees. In that case, we wouldn't want to mark it as executed but would be fine with it being retried next block.

Reply:
To do this we need to spawn another service which monitor message from l1 extrinsic execution, and insert the transaction again if needed. It won't just "reexecute itself".

Fees for l1->l2 messages are paid on l1. No other fee is deducted during execution on l1. So it shouldn't fail this way.

Reply:
Well at least it should be noted that the sequencer always needs to have enough ETH on l2 to cover message fees otherwise they will *never* be executed

Reply:
What? No.
No fee are taken for l1-handler txs.
```rust
        let paid_fee = self.paid_fee_on_l1;
        // For now, assert only that any amount of fee was paid.
        // The error message still indicates the required fee.
        if paid_fee == Fee(0) {
            return Err(TransactionFeeError::InsufficientL1Fee { paid_fee, actual_fee })?;
        }
```

That's the only fee-related thing is the code.
Nobody needs any ETH to pay for anything on l2.